### PR TITLE
Add an AWS S3 scan result storage

### PR DESCRIPTION
**Pull request by carlos-ardila2**
Add support for using AWS S3 as online cache for scan results.

Provides a S3FileStorage class which extends [FileStorage](https://github.com/oss-review-toolkit/ort/blob/main/utils/ort/src/main/kotlin/storage/FileStorage.kt) and uses official Amazon's  [AWS Java SDK](https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/examples-s3.html) to store scan results.

Issue oss-review-toolkit/ort#730

### CODE REVIEW

File: model/src/main/kotlin/config/S3FileStorageConfiguration.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (C) 2017 The ORT Project Authors (see <https://github.com/oss-review-toolkit/ort/blob/main/NOTICE>)
```
Comment (tsteenbe) [inline model/src/main/kotlin/config/S3FileStorageConfiguration.kt:2]: S3FileStorageConfiguration.kt is a new file so please change copyright year from 2017 to 2023

File: utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt [hunk 2]
```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright (C) 2019 The ORT Project Authors (see <https://github.com/oss-review-toolkit/ort/blob/main/NOTICE>)
```
Comment (tsteenbe) [inline utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt:2]: S3FileStorageFunTest.kt is a new file so please change copyright year from 2019 to 2023

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt [hunk 3]
```diff
@@ -0,0 +1,190 @@
+/*
+ * Copyright (C) 2019 The ORT Project Authors (see <https://github.com/oss-review-toolkit/ort/blob/main/NOTICE>)
```
Comment (tsteenbe) [inline utils/ort/src/main/kotlin/storage/S3FileStorage.kt:2]: S3FileStorage.kt is a new file so please change copyright year from 2019 to 2023

File: README.md [hunk 4]
```diff
@@ -546,6 +546,30 @@ ort:
     storageWriters: ["artifactoryStorage"]
 ```
 
+### AWS S3 Storage
+
+An AWS S3 Bucket can be used to store scan results. You must provide a previously created Bucket, an optional AWS region
```
Comment (tsteenbe) [inline README.md:551]: Could you make "AWS S3 Bucket" a link to https://docs.aws.amazon.com/AmazonS3/latest/userguide/Welcome.html? Will help ORT users quickly find the docs on how to set up a bucket.

File: README.md [hunk 5]
```diff
@@ -546,6 +546,30 @@ ort:
...
+(if not present it will use [default](https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/region-selection.html))
+and credentials (if not provided, system [default](https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/credential-providers.html)
+values will be used). Optionally, you can set the data to be stored compressed.
+
+```yaml
+ort:
+  scanner:
+    storages:
+      awsStorage:
```
Comment (tsteenbe) [inline README.md:560]: AWS offers multiple storage services see https://aws.amazon.com/products/storage/ therefore I believe it's better to not name the key `awsStorage` in case we want to add other AWS storage service in the future.

What do you think of renaming `awsStorage` to `awsS3Storage` across this PR?

**Review by tsteenbe — CHANGES_REQUESTED** [review]

File: gradle/libs.versions.toml [hunk 6]
```diff
@@ -42,12 +42,13 @@ logback = "1.4.11"
 maven = "3.9.4"
 mavenResolver = "1.9.15"
 mockk = "1.13.7"
-okhttp = "4.11.0"
+okhttp = "5.0.0-alpha.11"
```
Comment (sschuberth) [inline gradle/libs.versions.toml:45]: Why's this necessary? We shouldn't use alpha version dependencies in production code (and ORT's `main` *is* used in production, as it's supposed to always be stable).

File: gradle/libs.versions.toml [hunk 7]
```diff
@@ -42,12 +42,13 @@ logback = "1.4.11"
...
 postgres = "42.6.0"
 postgresEmbedded = "1.0.1"
 reflections = "0.10.2"
 retrofit = "2.9.0"
 retrofitConverterKotlinxSerialization = "1.0.0"
+s3 = "0.29.0-beta"
```
Comment (sschuberth) [inline gradle/libs.versions.toml:51]: Also here, please use the latest stable / non-beta version, if possible.

File: gradle/libs.versions.toml [hunk 8]
```diff
@@ -70,6 +71,7 @@ versions = { id = "com.github.ben-manes.versions", version.ref = "versionsPlugin
 antlr = { module = "org.antlr:antlr4", version.ref = "antlr" }
 asciidoctorj = { module = "org.asciidoctor:asciidoctorj", version.ref = "asciidoctorj" }
 asciidoctorjPdf = { module = "org.asciidoctor:asciidoctorj-pdf", version.ref = "asciidoctorjPdf" }
+aws-s3 = { module = "aws.sdk.kotlin:s3", version.ref = "s3" }
```
Comment (sschuberth) [inline gradle/libs.versions.toml:74]: Use CamelCase instead of snake-case for the key (see existing code).

File: model/src/main/kotlin/config/S3FileStorageConfiguration.kt [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+package org.ossreviewtoolkit.model.config
+
+/**
+ * A class to hold the configuration for using local files as a storage.
+ */
+data class S3FileStorageConfiguration(
+    /**
+     * The name of the S3 bucket used to store files in.
+     */
```
Comment (sschuberth) [inline model/src/main/kotlin/config/S3FileStorageConfiguration.kt:28]: You may want to use more compact one-liners like

    /** The name of the S3 bucket used to store files in. */

to document properties.

Also, consider adding empty lines to separate the properties, so it's visually more clear which comment belongs to which property.

File: model/src/test/kotlin/config/OrtConfigurationTest.kt [hunk 10]
```diff
@@ -229,6 +231,7 @@ class OrtConfigurationTest : WordSpec({
                         this shouldContainExactly mapOf(
                             "commandLine" to "--copyright --license --info --strip-root --timeout 300",
                             "commandLineNonConfig" to "--processes 4",
+                            "parseLicenseExpressions" to "true",
```
Comment (sschuberth) [inline model/src/test/kotlin/config/OrtConfigurationTest.kt]: This should be dropped.

File: utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+        handler.requests.clear()
+    }
+
+    override suspend fun afterSpec(spec: Spec) {
+        // Ensure the server is properly stopped even in case of exceptions, but wait at most 5 seconds.
+        server.stop(5)
+    }
+
+    init {
+
```
Comment (sschuberth) [inline utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt:122]: Please drop this superfluous empty line.

File: utils/ort/src/main/kotlin/OkHttpClientHelper.kt [hunk 12]
```diff
@@ -209,8 +209,8 @@ fun OkHttpClient.download(url: String, acceptEncoding: String? = null): Result<P
     }.mapCatching { response ->
         val body = response.body
 
-        if (!response.isSuccessful || body == null) {
-            body?.close()
+        if (!response.isSuccessful) {
+            body.close()
```
Comment (sschuberth) [inline utils/ort/src/main/kotlin/OkHttpClientHelper.kt:213]: Why this change? Does AWS allow null bodies on success? If so, this should be extracted to a preceding preparing refactoring commit, that thoroughly explains this in the commit message.

Comment (sschuberth) [reply utils/ort/src/main/kotlin/OkHttpClientHelper.kt:213]: Ah, wait, this is due to the OkHttp alpha upgrade, and thus should be dropped together with the upgrade.

File: utils/ort/src/main/kotlin/storage/HttpFileStorage.kt [hunk 13]
```diff
@@ -105,12 +105,9 @@ class HttpFileStorage(
 
         val response = httpClient.execute(request)
         if (response.isSuccessful) {
-            response.body?.let { body ->
+            response.body.let { body ->
```
Comment (sschuberth) [inline utils/ort/src/main/kotlin/storage/HttpFileStorage.kt:108]: Drop this, see above.

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+
+/**
+ * A [FileStorage] that stores files in an AWS S3 bucket [bucketName]. The [read] and [exists] operations are
+ * blocking, but the [write] operation is asynchronous unless a [customEndpoint] is provided.
+ */
+@Suppress("SwallowedException")
+class S3FileStorage(
+    /**
+     * The name of the S3 bucket used to store files in.
+     */
```
Comment (sschuberth) [inline utils/ort/src/main/kotlin/storage/S3FileStorage.kt:60]: Consider more compact comments / empty separator lines as outlined above.

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+        var fileExists = true
+
+        runBlocking {
+            try {
+                s3.headObject(request)
+            } catch (e: NotFound) {
+                fileExists = false
+            }
+        }
```
Comment (sschuberth) [inline utils/ort/src/main/kotlin/storage/S3FileStorage.kt:129]: Probably better something like

    return runBlocking { runCatching { s3.headObject(request) }.isSuccess }

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+
+        val stream: InputStream
+
+        runBlocking {
+            stream = job.await()
+        }
+
+        return stream
```
Comment (sschuberth) [inline utils/ort/src/main/kotlin/storage/S3FileStorage.kt:157]: `return runBlocking { job.await() }`

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+                 PutObjectRequest {
+                    key = path
+                    bucket = bucketName
+                    body = ByteStream.fromBytes(it.readBytes())
+                }
+            }.let {
+                try {
+                    s3.putObject(it)
+                } catch (e: S3Exception) {
+                    logger.warn("Can not write $path to S3 bucket $bucketName")
```
Comment (sschuberth) [inline utils/ort/src/main/kotlin/storage/S3FileStorage.kt:178]: Use lazy curly-brace syntax with `log`-statements: `log.warn {}`

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+                }
+            }
+        }
+
+        // Local server doesn't play nice with auto scheduled coroutines. Run blocking instead
```
Comment (sschuberth) [inline utils/ort/src/main/kotlin/storage/S3FileStorage.kt:183]: Please properly end (even comment) sentences with a dot, and generally use proper grammar like in a written user's manual, not informal language.

File: model/src/main/kotlin/config/S3FileStorageConfiguration.kt [hunk 19]
```diff
@@ -30,13 +30,17 @@ data class S3FileStorageConfiguration(
...
     /**
      * The AWS access key.
      */
-     val accessKeyId: String?,
+    val accessKeyId: String?,
     /**
      * The AWS secret for the access key.
      */
-     val secretAccessKey: String?
+    val secretAccessKey: String?,
```
Comment (sschuberth) [inline model/src/main/kotlin/config/S3FileStorageConfiguration.kt]: These lines should be amended to the previous commit.

File: README.md [hunk 20]
```diff
@@ -551,7 +551,7 @@ ort:
 An AWS S3 Bucket can be used to store scan results. You must provide a previously created Bucket, an optional AWS region
 (if not present it will use [default](https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/region-selection.html)) 
 and credentials (if not provided, system [default](https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/credential-providers.html) 
-values will be used).
+values will be used). Optionally, you can set the data to be stored compressed.
```
Comment (sschuberth) [inline README.md]: The commit message should probably also mention that compression is optional.

File: README.md [hunk 21]
```diff
@@ -548,16 +548,18 @@ ort:
 
 ### AWS S3 Storage
 
-An AWS S3 Bucket can be used to store scan results. You must provide a previously created Bucket, an optional AWS region
-(if not present it will use [default](https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/region-selection.html))
-and credentials (if not provided, system [default](https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/credential-providers.html)
-values will be used). Optionally, you can set the data to be stored compressed.
+An [AWS S3 Bucket](https://docs.aws.amazon.com/AmazonS3/latest/userguide/Welcome.html) can be used to store scan
```
Comment (sschuberth) [inline README.md:551]: Again, move the hunks in this commit to the commits that introduced the original respective code.

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Comment by sschuberth** [comment]
@carlos-ardila2 is there any interest from your side to move this forward?

File: gradle/libs.versions.toml (hunk 7 above)
Comment (carlos-ardila2) [reply gradle/libs.versions.toml:51]: [0.32.4-beta](https://mvnrepository.com/artifact/aws.sdk.kotlin/s3/0.32.4-beta) is the latest version of the official Kotlin SDK provided by Amazon. We could try to use their Java SDK, [2.20.162](https://mvnrepository.com/artifact/software.amazon.awssdk/s3/2.20.162) But I believe it will defy the purpose of having the best version available from the vendor, which uses actual Kotlin features.

Comment (sschuberth) [reply gradle/libs.versions.toml:51]: Ok, I see, the AWS SDK for Kotlin is a developer preview, and there hasn't been any non-beta release since version [0.16.0](https://mvnrepository.com/artifact/aws.sdk.kotlin/s3/0.16.0) from 2022. So let's stick to the latest beta version then.

File: gradle/libs.versions.toml (hunk 6 above)
Comment (carlos-ardila2) [reply gradle/libs.versions.toml:45]: Amazon has [stated](https://github.com/awslabs/aws-sdk-kotlin/issues/765#issuecomment-1374093175) that their Koltin library relies on the latest version of okhttp, which oddly enough is the alpha one.  I'm not sure how to proceed.

Comment (sschuberth) [reply gradle/libs.versions.toml:45]: > which oddly enough is the alpha one.

Thanks for this information, indeed odd. I don't think it's acceptable for us to depend on an alpha version of OkHttp for the whole of ORT. We could solve this if 1) [scan storages were plugin](https://github.com/oss-review-toolkit/ort/issues/6603) and 2) [plugin classpaths were isolated from ORT classpath](https://github.com/oss-review-toolkit/ort/issues/6798), but both are bigger refactorings that I do not see happening any time soon.

> I'm not sure how to proceed.

I guess our best option then is to indeed use the Java instead of the Kotlin flavor of the AWS SDK...

File: README.md (hunk 21 above)
Comment (carlos-ardila2) [reply README.md:551]: Addressed by squashing

File: integrations/schemas/ort-configuration-schema.json [hunk 22]
```diff
@@ -540,6 +540,7 @@
         },
         "StorageTypes": {
             "enum": [
+                "aws",
```
[the remark this answers is not in the document]
Comment (carlos-ardila2) [reply integrations/schemas/ort-configuration-schema.json:543]: Addressed by squashing

File: gradle/libs.versions.toml (hunk 6 above)
Comment (carlos-ardila2) [reply gradle/libs.versions.toml:47]: I'll start working on a refactor.

File: model/src/main/kotlin/config/S3FileStorageConfiguration.kt (hunk 1 above)
Comment (carlos-ardila2) [reply model/src/main/kotlin/config/S3FileStorageConfiguration.kt]: Addressed by squashing (not the best way to do it, I agree)

File: utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt (hunk 2 above)
Comment (carlos-ardila2) [reply utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt]: Addressed by squashing

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt (hunk 3 above)
Comment (carlos-ardila2) [reply utils/ort/src/main/kotlin/storage/S3FileStorage.kt]: Addressed by squashing

File: gradle/libs.versions.toml (hunk 7 above)
Comment (carlos-ardila2) [reply gradle/libs.versions.toml]: Using the latest now. Although not important now that we are refactoring to use the java library

File: gradle/libs.versions.toml (hunk 8 above)
Comment (carlos-ardila2) [reply gradle/libs.versions.toml]: Fixed, thank you

File: model/src/main/kotlin/config/S3FileStorageConfiguration.kt (hunk 9 above)
Comment (carlos-ardila2) [reply model/src/main/kotlin/config/S3FileStorageConfiguration.kt]: Addressed. Although, there are many other files in the codebase that still use three lines of comment per field.

File: model/src/main/kotlin/config/S3FileStorageConfiguration.kt (hunk 19 above)
Comment (carlos-ardila2) [reply model/src/main/kotlin/config/S3FileStorageConfiguration.kt]: Addressed by squashing.

File: README.md (hunk 20 above)
Comment (carlos-ardila2) [reply README.md]: Addressed by squashing.

File: README.md [hunk 23]
```diff
@@ -385,7 +385,7 @@ supported:
   * [PNPM](https://pnpm.io/) (limitations:
     [no peer dependencies](https://github.com/oss-review-toolkit/ort/issues/95))
   * [Yarn 1](https://classic.yarnpkg.com/)
-  * [Yarn 2+](https://next.yarnpkg.com/)
+  * [Yarn 2+](https://yarnpkg.com/)
```
[the remark this answers is not in the document]
Comment (carlos-ardila2) [reply README.md]: Addressed by squashing.

[the remark this answers is not in the document]
Comment (carlos-ardila2) [reply README.md]: Addressed by squashing.

File: model/src/test/kotlin/config/OrtConfigurationTest.kt (hunk 10 above)
Comment (carlos-ardila2) [reply model/src/test/kotlin/config/OrtConfigurationTest.kt]: Addressed by squashing.

File: utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt (hunk 11 above)
Comment (carlos-ardila2) [reply utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt]: Addressed

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt (hunk 14 above)
Comment (carlos-ardila2) [reply utils/ort/src/main/kotlin/storage/S3FileStorage.kt]: Addressed

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt (hunk 15 above)
Comment (carlos-ardila2) [reply utils/ort/src/main/kotlin/storage/S3FileStorage.kt]: Thank you for that valuable suggestion. I'd still like to keep some record if the expection is any different form NotFound.

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt (hunk 17 above)
Comment (carlos-ardila2) [reply utils/ort/src/main/kotlin/storage/S3FileStorage.kt]: Addressed

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt (hunk 18 above)
Comment (carlos-ardila2) [reply utils/ort/src/main/kotlin/storage/S3FileStorage.kt]: Addressed

**Comment by carlos-ardila2** [comment]
@sschuberth I'm back in. I apologize for taking this long to get back to this task. The AWS requirement to use an alpha version for the Kotlin S3 library was a set-back. I've addressed most of the issues and squash in a single commit. I know it is not the best way to do it, but I thought it would be best to start fresh for the refactor. Regards.

**Comment by carlos-ardila2** [comment]
@sschuberth I'm back in. I apologize for taking this long to get back to this task. The AWS requirement to use an alpha version for the Kotlin S3 library was a set-back. I've addressed most of the issues and squash in a single commit. I know it is not the best way to do it, but I thought it would be best to start fresh for the refactor. Regards.

File: gradle/libs.versions.toml (hunk 6 above)
Comment (carlos-ardila2) [reply gradle/libs.versions.toml]: I have completed the refactor, please review.

File: utils/ort/src/main/kotlin/storage/HttpFileStorage.kt (hunk 13 above)
Comment (carlos-ardila2) [reply utils/ort/src/main/kotlin/storage/HttpFileStorage.kt]: Fixed with refactor.

File: utils/ort/src/main/kotlin/OkHttpClientHelper.kt (hunk 12 above)
Comment (carlos-ardila2) [reply utils/ort/src/main/kotlin/OkHttpClientHelper.kt]: Fixed with refactor.

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt (hunk 16 above)
Comment (carlos-ardila2) [reply utils/ort/src/main/kotlin/storage/S3FileStorage.kt]: No longer part of the code base after refactor.

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt [hunk 24]
```diff
@@ -0,0 +1,173 @@
...
+                    val stream = ByteArrayInputStream(resp.body?.toByteArray())
+                    if (compression) XZCompressorInputStream(stream) else stream
+                }
+            } catch (e: NoSuchKey) {
+                throw NoSuchFileException(File(path))
+            }
+        }
+
+        return runBlocking { job.await() }
+
```
Comment (sschuberth) [inline utils/ort/src/main/kotlin/storage/S3FileStorage.kt]: Please drop this trailing blank line.

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: gradle/libs.versions.toml (hunk 6 above)
Comment (carlos-ardila2) [reply gradle/libs.versions.toml]: I'm sorry, I'm afraid I need some clarification. There are two commits on the branch right now, the first one contains the version that uses the experimental Kotlin library by Amazon. Then another commit with a refactor switching to Java library, therefore, this alpha version (and other changes mentioned below) are found **solely** in the first commit but **not** in the HEAD. Should this first commit be squashed too? What I usually do is to keep commits until just before merging, then I squash them, just in case I need to track down errors and even discarded ideas. Could you please explain me this project's policy regarding that?

File: utils/ort/src/main/kotlin/OkHttpClientHelper.kt (hunk 12 above)
Comment (carlos-ardila2) [reply utils/ort/src/main/kotlin/OkHttpClientHelper.kt]: Only in the first commit (won't compile otherwise), gone in the HEAD.

File: utils/ort/src/main/kotlin/storage/HttpFileStorage.kt (hunk 13 above)
Comment (carlos-ardila2) [reply utils/ort/src/main/kotlin/storage/HttpFileStorage.kt]: Only in the first commit (won't compile otherwise), gone in the HEAD.

File: gradle/libs.versions.toml (hunk 6 above)
Comment (sschuberth) [reply gradle/libs.versions.toml]: > There are two commits on the branch right now, the first one contains the version that uses the experimental Kotlin library by Amazon.

Ah, my bad, I might have accidentally looked only at the first commit, assuming there would be only one / believing to look at the combined diff.

> Should this first commit be squashed too?

Yes please. We don't squash upon merge (as that combines individual commits from a PR), but rebase upon merge (also merge commit are not allowed).

> What I usually do is to keep commits until just before merging, then I squash them, just in case I need to track down errors and even discarded ideas.

Makes sense, but locally you could probably use `git reflog` to restore intermediate states.

Comment (carlos-ardila2) [reply gradle/libs.versions.toml]: Got it! Thans for the `git reflog ` tip, life saver.

**Comment by carlos-ardila2** [comment]
> @carlos-ardila2 please rebase this once more onto latest `main` to get rid of some unrelated code changes ("afterwards" -> "afterward") and to address unrelated test failures.

Done

File: README.md [hunk 25]
```diff
@@ -546,6 +546,32 @@ ort:
...
+(if not provided, system
+[default](https://docs.aws.amazon.com/sdk-for-kotlin/latest/developer-guide/credential-providers.html) values will be
+used). Optionally, you can set the data to be compressed before stored.
+
+```yaml
+ort:
+  scanner:
+    storages:
+      awsS3Storage:
+        backend:
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply README.md:565]: Ping @carlos-ardila2, please see above.

File: utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt [hunk 26]
```diff
@@ -0,0 +1,152 @@
...
+import java.net.InetAddress
+import java.net.InetSocketAddress
+import java.util.Collections
+
+import kotlin.random.Random
+
+import org.apache.logging.log4j.kotlin.logger
+
+class S3FileStorageFunTest : WordSpec() {
+
```
Comment (sschuberth) [inline utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt:44]: Nit: Please drop this superfluous blank line.

File: utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt [hunk 27]
```diff
@@ -0,0 +1,152 @@
...
+    private val loopback = InetAddress.getLoopbackAddress()
+    private val protocol = "http"
+    private val port = Random.nextInt(1024, 49152) // See https://en.wikipedia.org/wiki/Registered_port.
+        .also { logger.debug { "Using port $it for S3 Mock server." } }
+
+    private val bucket = "ort-scan-results"
+
+    private val handler = object : HttpHandler {
+        val requests = Collections.synchronizedMap(mutableMapOf<String, String>())
```
Comment (sschuberth) [inline utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt:53]: As tests within a single spec are not run in parallel, can we stick to a simple `mutableMapOf<String, String>()` like in `HttpFileStorageFunTest`?

File: utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt [hunk 28]
```diff
@@ -0,0 +1,152 @@
...
+                    exchange.sendResponseHeaders(status, -1)
+                }
+                "PUT" -> {
+                    val key = exchange.requestURI.toString().removePrefix("/$bucket/")
+                    requests[key] = exchange.requestBody.reader().use { it.readText() }.split("\n")[1].trimEnd()
+                    exchange.sendResponseHeaders(HttpURLConnection.HTTP_OK, 0)
+                }
+                "GET" -> {
+                    val key = exchange.requestURI.toString().removePrefix("/$bucket/")
+                    requests[key].let { data ->
```
Comment (sschuberth) [inline utils/ort/src/funTest/kotlin/storage/S3FileStorageFunTest.kt:72]: There doe snot seem to be a benefit by using `.let` here, so consider using plain `val data = requests[key]` to reduce nesting of following code.

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt [hunk 29]
```diff
@@ -0,0 +1,146 @@
...
+import software.amazon.awssdk.services.s3.model.S3Exception
+
+/**
+ * A [FileStorage] that stores files in an AWS S3 bucket [bucketName]. The [read] and [exists] operations are
+ * blocking, but the [write] operation is asynchronous unless a [customEndpoint] is provided. Contents are compressed
+ * before store if the [compression] flag is set to true.
+ */
+@Suppress("SwallowedException")
+class S3FileStorage(
+
```
Comment (sschuberth) [inline utils/ort/src/main/kotlin/storage/S3FileStorage.kt:50]: Nit: Please drop this superfluous blank line.

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt [hunk 30]
```diff
@@ -0,0 +1,146 @@
...
+    private val compression: Boolean = false,
+
+    /** Custom endpoint to perform AWS API Requests */
+    private val customEndpoint: String? = null,
+
+    /** The AWS secret for the access key. */
+    private val secretAccessKey: String? = null
+
+) : FileStorage {
+
```
Comment (sschuberth) [inline utils/ort/src/main/kotlin/storage/S3FileStorage.kt:70]: Nit: Please drop this superfluous blank line.

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt [hunk 31]
```diff
@@ -0,0 +1,146 @@
...
+    /** Whether to use compression for storing files or not. Defaults to true. */
+    private val compression: Boolean = false,
+
+    /** Custom endpoint to perform AWS API Requests */
+    private val customEndpoint: String? = null,
+
+    /** The AWS secret for the access key. */
+    private val secretAccessKey: String? = null
+
```
Comment (sschuberth) [inline utils/ort/src/main/kotlin/storage/S3FileStorage.kt:68]: Nit: Please drop this superfluous blank line.

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt [hunk 32]
```diff
@@ -0,0 +1,146 @@
...
+) : FileStorage {
+
+    val s3: S3Client? by lazy {
+        val provider = if (accessKeyId != null && secretAccessKey != null) {
+            StaticCredentialsProvider.create(AwsBasicCredentials.create(accessKeyId, secretAccessKey))
+        } else {
+            null
+        }
+
+        val client = if (awsRegion != null && provider != null) {
```
Comment (sschuberth) [inline utils/ort/src/main/kotlin/storage/S3FileStorage.kt:78]: Just omit the `val client = ` here and use `if` as an expression that returns the value.

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt [hunk 33]
```diff
@@ -0,0 +1,146 @@
...
+            }.build()
+        } else {
+            if (awsRegion != null) {
+                S3Client.builder().region(Region.of(awsRegion)).build()
+            } else {
+                S3Client.create()
+            }
+        }
+
+        client
```
Comment (sschuberth) [inline utils/ort/src/main/kotlin/storage/S3FileStorage.kt:92]: This can be dropped when using `if` as an expression, see above.

File: utils/ort/src/main/kotlin/storage/S3FileStorage.kt [hunk 34]
```diff
@@ -0,0 +1,146 @@
...
+
+    /** Custom endpoint to perform AWS API Requests */
+    private val customEndpoint: String? = null,
+
+    /** The AWS secret for the access key. */
+    private val secretAccessKey: String? = null
+
+) : FileStorage {
+
+    val s3: S3Client? by lazy {
```
Comment (sschuberth) [inline utils/ort/src/main/kotlin/storage/S3FileStorage.kt:71]: Let's name the variable `s3Client` for clarity.

Comment (sschuberth) [reply utils/ort/src/main/kotlin/storage/S3FileStorage.kt:71]: Also, can this really ever be null? I acknowledge this comes from a Java platform call that doesn't support explicit non-nullability. But according to the docs the builder should never return `null`, or?

**Review by sschuberth — CHANGES_REQUESTED** [review]

**Review by sschuberth — APPROVED** [review]

**Comment by sschuberth** [comment]
Test failures are unrelated.
